File: web/src/main/java/be/ucll/runetracker/database/RunescapeUserDatabaseJPA.java
package be.ucll.runetracker.database;

import be.ucll.runetracker.domain.RunescapeUser;

import javax.persistence.*;
import java.util.Collection;
import java.util.Optional;
import java.util.stream.Stream;

public class RunescapeUserDatabaseJPA implements RunescapeUserDatabase {
    private final EntityManager entityManager;

    public RunescapeUserDatabaseJPA() {
        EntityManagerFactory factory = Persistence.createEntityManagerFactory(
                "RuneTrackerPU"
        );
        entityManager = factory.createEntityManager();
    }

    @Override
    public void add(RunescapeUser RunescapeUser) {
        EntityTransaction transaction = entityManager.getTransaction();
        transaction.begin();
        try {
            entityManager.persist(RunescapeUser);
            transaction.commit();
        } catch (Exception e) {
            transaction.rollback();
        }
    }

    @Override
    public void addAll(Collection<RunescapeUser> users) {
        EntityTransaction transaction = entityManager.getTransaction();
        transaction.begin();
        try {
            entityManager.setFlushMode(FlushModeType.COMMIT);
            for (RunescapeUser RunescapeUser : users) {
                entityManager.persist(RunescapeUser);
            }
            transaction.commit();
        } catch (Exception e) {
            transaction.rollback();
        } finally {
            entityManager.setFlushMode(FlushModeType.AUTO);
        }
    }

    @Override
    public void delete(RunescapeUser RunescapeUser) {
        EntityTransaction transaction = entityManager.getTransaction();
        transaction.begin();
        try {
            entityManager.remove(RunescapeUser);
            transaction.commit();
        } catch (Exception e) {
            transaction.rollback();
        }
    }

    @Override
    public Optional<RunescapeUser> get(int id) {
        return Optional.ofNullable(entityManager.find(RunescapeUser.class, id));
    }

    @Override
    public Stream<RunescapeUser> all() {
        return entityManager.createQuery("SELECT u FROM RunescapeUser u", RunescapeUser.class).getResultStream();
    }

    @Override
    public void update(RunescapeUser user) {
        EntityTransaction transaction = entityManager.getTransaction();
        transaction.begin();
        try {
            entityManager.merge(user);
            transaction.commit();
        } catch (Exception e) {
            transaction.rollback();
        }
    }
}

File: web/src/main/java/me/robinvdb/runetracker/database/RunescapeUserDatabaseJPA.kt
package me.robinvdb.runetracker.database

import me.robinvdb.runetracker.domain.RunescapeUser

import javax.persistence.*
import java.util.Optional
import java.util.stream.Stream

class RunescapeUserDatabaseJPA : RunescapeUserDatabase {
    private val entityManager: EntityManager

    init {
        val factory = Persistence.createEntityManagerFactory(
                "RuneTrackerPU"
        )
        entityManager = factory.createEntityManager()
    }

    override fun add(user: RunescapeUser) {
        val transaction = entityManager.transaction
        transaction.begin()
        try {
            entityManager.persist(user)
            transaction.commit()
        } catch (e: Exception) {
            transaction.rollback()
        }

    }

    override fun addAll(users: Collection<RunescapeUser>) {
        val transaction = entityManager.transaction
        transaction.begin()
        try {
            entityManager.flushMode = FlushModeType.COMMIT
            for (RunescapeUser in users) {
                entityManager.persist(RunescapeUser)
            }
            transaction.commit()
        } catch (e: Exception) {
            transaction.rollback()
        } finally {
            entityManager.flushMode = FlushModeType.AUTO
        }
    }

    override fun delete(user: RunescapeUser) {
        val transaction = entityManager.transaction
        transaction.begin()
        try {
            entityManager.remove(user)
            transaction.commit()
        } catch (e: Exception) {
            transaction.rollback()
        }

    }

    override fun get(id: Int): Optional<RunescapeUser> {
        return Optional.ofNullable(entityManager.find(RunescapeUser::class.java, id))
    }

    override fun all(): Stream<RunescapeUser> {
        return entityManager.createQuery("SELECT u FROM RunescapeUser u", RunescapeUser::class.java).resultStream
    }

    override fun update(user: RunescapeUser) {
        val transaction = entityManager.transaction
        transaction.begin()
        try {
            entityManager.merge(user)
            transaction.commit()
        } catch (e: Exception) {
            transaction.rollback()
        }

    }
}
